README.md

```diff
@@ -1,2 +1,193 @@
 # ast-gh-actions
+
 A library or collection of AST github shared composite actions.
+
+## Development
+
+### Git Pre-commit Hook
+
+The repository uses a pre-commit hook to run linting on all TypeScript projects. To enable:
+
+```bash
+git config core.hooksPath .githooks
+```
+
+This runs automatically when installing dependencies in the root directory.
+
+When adding a new action, update `.githooks/pre-commit` to include your project:
+
+```bash
+#!/bin/bash
+# set -e exits on error, making `|| exit 1` redundant
+set -euo
+
+# Print each command before executing
+set -x
+
+for project in upstream-tag-on-merge upstream-tag-sync my-new-action; do
+    cd $project
+    echo "Building $project..."
+    bun run build
+    git add dist
+    echo "Linting $project..."
+    bun run lint
+    cd ..
+done
+```
+
+The pre-commit hook:
+
+- Uses bash safety features (`set -euo`)
+- Prints commands for debugging (`set -x`)
+- Builds TypeScript files
+- Stages compiled dist files
+- Runs linting
+
+### GitHub Workflows
+
+#### Lint
+
+`.github/workflows/lint.yml` runs on PR creation and updates:
+
+- Runs ESLint on all TypeScript files
+- Executes in parallel for each project
+- Fails if linting errors are found
+
+#### Build Verification
+
+`.github/workflows/verify-build.yml` ensures compiled code matches source:
+
+- Cleans and rebuilds TypeScript
+- Compares with committed build artifacts
+- Fails if builds are out of sync
+
+## Adding New Actions
+
+When creating a new shared GitHub Action:
+
+1. Create a new directory for your action:
+
+```bash
+mkdir my-new-action
+cd my-new-action
+```
+
+2. Initialize the project:
+
+```bash
+bun init
+```
+
+3. Add to GitHub workflows by updating the matrix in both workflow files:
+
+```yaml
+strategy:
+  matrix:
+    project: ['upstream-tag-on-merge', 'upstream-tag-sync', 'my-new-action']
+```
+
+4. Required files:
+
+- `src/index.ts` - Action source code
+- `action.yml` - Action metadata
+- `package.json` - Dependencies and scripts
+- `tsconfig.json` - TypeScript configuration
+
+5. Required files and configurations:
```

Comment:
I'm not too familiar with GitHub Actions repositories but if we want to have the same linting/setup/build configuration everywhere and unless I missed something, shouldn't we centralize them rather than duplicating them?

\+ please point to the files themselves rather than keeping here a copy that may get outdated.

Reply:
Maybe we should get rid of these section ` Required files and configurations` and do not require any specific config files but rather let the way linting, formating etc is implemented to be decided by the specific action. I think this might go a bit into what @RudolfSchreier also already mentioned [here](https://github.com/swisstxt/ast-gh-actions/pull/5#discussion_r1932090898). Therefore we might just want to define that a `lint` script exists in the `package.json` but all the complexity about how exactly things are linted can be made up by the specific action. For example the action `upstream-tag-on-merge` might define some specific `eslint.config.js` properties which make only sense there and therefore this also can't be centralized. @devnoname120 @RudolfSchreier What do you think?

Reply:
@domcyrus Sounds good to me, we can revisit later if needed